e2e/e2etests/test_precompiles_distribute.go

```diff
@@ -0,0 +1,196 @@
+	utils.RequiredTxFailed(r, receipt, "distribute should fail when there's no allowance")
+
+	// Balances shouldn't change after a failed attempt.
+	balanceShouldBe(r, 1000, checkZRC20Balance(r, spenderAddress))
+	balanceShouldBe(r, 0, checkZRC20Balance(r, lockerAddress))
+	balanceShouldBe(r, 0, checkCosmosBalance(r, feeCollectorAddress, zrc20Denom))
+
+	// Allow 500.
+	approveAllowance(r, distributeContractAddress, fiveHundred)
+
+	// Shouldn't be able to distribute more than allowed.
+	tx, err = dstrContract.Distribute(r.ZEVMAuth, zrc20Address, fiveHundredOne)
+	require.NoError(r, err)
+	receipt = utils.MustWaitForTxReceipt(r.Ctx, r.ZEVMClient, tx, r.Logger, r.ReceiptTimeout)
+	utils.RequiredTxFailed(r, receipt, "distribute should fail trying to distribute more than allowed")
+
+	// Balances shouldn't change after a failed attempt.
+	balanceShouldBe(r, 1000, checkZRC20Balance(r, spenderAddress))
+	balanceShouldBe(r, 0, checkZRC20Balance(r, lockerAddress))
+	balanceShouldBe(r, 0, checkCosmosBalance(r, feeCollectorAddress, zrc20Denom))
+
+	// Raise the allowance to 1000.
+	approveAllowance(r, distributeContractAddress, oneThousand)
+
+	// Shouldn't be able to distribute more than owned balance.
+	tx, err = dstrContract.Distribute(r.ZEVMAuth, zrc20Address, oneThousandOne)
+	require.NoError(r, err)
+	receipt = utils.MustWaitForTxReceipt(r.Ctx, r.ZEVMClient, tx, r.Logger, r.ReceiptTimeout)
+	utils.RequiredTxFailed(r, receipt, "distribute should fail trying to distribute more than owned balance")
+
+	// Balances shouldn't change after a failed attempt.
+	balanceShouldBe(r, 1000, checkZRC20Balance(r, spenderAddress))
+	balanceShouldBe(r, 0, checkZRC20Balance(r, lockerAddress))
+	balanceShouldBe(r, 0, checkCosmosBalance(r, feeCollectorAddress, zrc20Denom))
+
+	// Should be able to distribute 500, which is within balance and allowance.
+	tx, err = dstrContract.Distribute(r.ZEVMAuth, zrc20Address, fiveHundred)
+	require.NoError(r, err)
+	receipt = utils.MustWaitForTxReceipt(r.Ctx, r.ZEVMClient, tx, r.Logger, r.ReceiptTimeout)
+	utils.RequireTxSuccessful(r, receipt, "distribute should succeed when distributing within balance and allowance")
+
+	balanceShouldBe(r, 500, checkZRC20Balance(r, spenderAddress))
+	balanceShouldBe(r, 500, checkZRC20Balance(r, lockerAddress))
+	balanceShouldBe(r, 500, checkCosmosBalance(r, feeCollectorAddress, zrc20Denom))
+
+	eventDitributed, err := dstrContract.ParseDistributed(*receipt.Logs[0])
+	require.NoError(r, err)
+	require.Equal(r, zrc20Address, eventDitributed.Zrc20Token)
+	require.Equal(r, spenderAddress, eventDitributed.Zrc20Distributor)
+	require.Equal(r, fiveHundred.Uint64(), eventDitributed.Amount.Uint64())
+
+	// After one block the rewards should have been distributed and fee collector should have 0 ZRC20 balance.
+	r.WaitForBlocks(1)
+	balanceShouldBe(r, 0, checkCosmosBalance(r, feeCollectorAddress, zrc20Denom))
+
+	// DO NOT REMOVE THE FOLLOWING CODE
```

Comment:
Do we have an issue to track it?

Reply:
No, I think there's no need since I'll start working immediately on this after merging the distribute function.

Reply:
Still better to create an issue so it's easy to visualize the steps in the epic